### PR TITLE
Node Task - Creating APIs using express  

### PR DESCRIPTION
**Pull request by ibratnawaz**
- Using express created APIs for the users.
- Storing the data in PostgreSQL DB.
- Used typescript.

### CODE REVIEW

**Comment by ibratnawaz** [comment]
@suraj-pareek, please review it. Thank you

File: src/middlewares/authenticate.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+export const checkDuplicateEmail = async (
+  req: Request,
+  res: Response,
+  next: NextFunction
+) => {
+  try {
+    const user = await User.count({ where: { login: req.body.login } });
+    if (user) {
+      res.status(400).json({ error: "Email is already taken." });
```
Comment (suraj-pareek) [inline src/middlewares/authenticate.ts]: It's a good practice to keep all HTTP status codes in a separate config/constant module.

File: src/models/user.ts [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        },
+      },
+    },
+    lastName: {
+      type: DataTypes.STRING,
+      allowNull: false,
+      validate: {
+        isAlpha: { msg: "Please prove valid first name." },
+        notEmpty: {
+          msg: "Fist name cannot be empty.",
```
Comment (suraj-pareek) [inline src/models/user.ts]: please correct the typo error.

File: src/middlewares/validate.ts [hunk 3]
```diff
@@ -0,0 +1,36 @@
+import { schema } from "../utils/joiValidation";
+import { Request, Response, NextFunction } from "express";
+
+export const payloadValidation = (
+  req: Request,
+  res: Response,
+  next: NextFunction
+) => {
+  let nextStep = true;
+  const keys = ["firstName", "lastName", "login", "password", "age"];
```
Comment (suraj-pareek) [inline src/middlewares/validate.ts]: please make it generic to handle all collections.

File: src/routes/userRoutes.ts [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+  checkUserPresent,
+  checkUserWithSoftDelete,
+} from "../middlewares/authenticate";
+import { payloadValidation } from "../middlewares/validate";
+
+const router: Router = express.Router();
+
+router
+  .route("/")
+  .post([payloadValidation, checkDuplicateEmail], createUser)
```
Comment (suraj-pareek) [inline src/routes/userRoutes.ts]: please give the appropriate middleware and controller name.

**Review by suraj-pareek — CHANGES_REQUESTED** [review]
Good work!

**Comment by ibratnawaz** [comment]
@suraj-pareek Please review it. Thank you

File: src/routes/userRoutes.ts [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+ */
+router
+  .route("/:id")
+  .put(authenticate({ paranoid: true }), updateUser)
+  .get(authenticate({ paranoid: true }), getUserById)
+  .delete(authenticate({ paranoid: true }), deleteUser);
+
+router.get("/restore/:id", authenticate({ paranoid: false }), restoreUser);
+
+router.get("/soft-deletes/list", getSoftDeltedUsers);
```
Comment (suraj-pareek) [inline src/routes/userRoutes.ts]: please correct endpoint;

File: src/models/user.ts [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+        notEmpty: {
+          msg: "First name cannot be empty.",
+        },
+      },
+    },
+    lastName: {
+      type: DataTypes.STRING,
+      allowNull: false,
+      validate: {
+        isAlpha: { msg: "Please prove valid first name." },
```
Comment (suraj-pareek) [inline src/models/user.ts]: please change the message as per fields

File: src/middlewares/validate.ts [hunk 7]
```diff
@@ -0,0 +1,38 @@
+import { schema } from "./users.joiValidation";
+import { Request, Response, NextFunction } from "express";
+import { NOT_ACCEPTABLE } from "../constants/statusCode";
+
+export const validation = (User: { rawAttributes: {} }) => {
```
Comment (suraj-pareek) [inline src/middlewares/validate.ts]: Please rename it to user validation as still its not generic validation. If you want to make it generic validation then you would get the columns/fields from the given schema and iterate over its properties.

File: src/routes/userRoutes.ts [hunk 8]
```diff
@@ -0,0 +1,36 @@
+import express, { Router } from "express";
+import {
+  getUsers,
+  createUser,
+  getUserById,
+  updateUser,
+  deleteUser,
+  restoreUser,
+  getSoftDeltedUsers,
```
Comment (suraj-pareek) [inline src/routes/userRoutes.ts]: can we have actives/inactive users instead of soft delete?

**Review by suraj-pareek — CHANGES_REQUESTED** [review]

**Comment by ibratnawaz** [comment]
@suraj-pareek Please review it. Thank you

**Comment by ibratnawaz** [comment]
Hi Suraj,
I have renamed the end-points and functions name as per you suggestion

**Review by suraj-pareek — APPROVED** [review]
Good Work!

**Pull request closed without merging**